prusti-specs/src/specifications/preparser.rs

```diff
@@ -289,557 +58,547 @@ impl Parse for SpecEntArgs {
-    expected_only_operator: bool,
-    /// A flag to denote that the parser is currently parsing a pledge
-    /// containing a lhs. This is important so that the parser stops at the
-    /// comma in between lhs and rhs.
-    parsing_pledge_with_lhs: bool,
+    /// Tokens yet to be consumed
+    tokens: VecDeque<TokenTree>,
+    /// Span of the last seen token
+    last_span: Option<Span>,
+    /// Span of the surrounding token
+    source_span: Option<Span>,
 }
 
 impl Parser {
+    /// initializes the parser with a TokenStream
     pub fn from_token_stream(tokens: TokenStream) -> Self {
-        let input = ParserStream::from_token_stream(tokens);
         Self {
-            input,
-            conjuncts: Vec::new(),
-            expr: Vec::new(),
-            previous_expression_resolved: false,
-            expected_operator: false,
-            expected_only_operator: false,
-            parsing_pledge_with_lhs: false,
+            tokens: tokens.into_iter().collect(),
+            last_span: None,
+            source_span: None,
         }
     }
-    fn from_parser_stream(input: ParserStream) -> Self {
+    /// initializes the parser with a TokenStream and the last span
+    fn from_token_stream_and_span(tokens: TokenStream, span: Option<Span>) -> Self {
         Self {
-            input,
-            conjuncts: Vec::new(),
-            expr: Vec::new(),
-            previous_expression_resolved: false,
-            expected_operator: false,
-            expected_only_operator: false,
-            parsing_pledge_with_lhs: false,
+            tokens: tokens.into_iter().collect(),
+            last_span: None,
+            source_span: span,
         }
     }
-    fn resolve_and(&mut self) -> syn::Result<()>{
-        // handles the case when there is no lhs of the && operator
-        if !self.expected_operator {
-            return Err(self.error_expected_assertion());
-        }
-
-        // convert the currently being parsed expression into a conjunct if not already
-        // done so
-        if self.previous_expression_resolved {
-            self.previous_expression_resolved = false;
-        }
-        else {
-            if self.expr.is_empty() {
-                return Err(self.error_expected_assertion());
+    /// Creates a single Prusti assertion from the input and returns it.
+    pub fn extract_assertion(&mut self) -> syn::Result<AssertionWithoutId> {
+        if self.tokens.is_empty() {
+            Ok(AssertionWithoutId{
+                kind: box common::AssertionKind::And(vec![])
+            })
+        } else {
+            if let Some(span) = self.contains_both_and_or(&self.tokens) {
+                return Err(self.error_ambiguous_expression(span));
             }
-            if let Err(err) = self.convert_expr_into_conjunct() {
-                return Err(err);
+
+            let expr = self.parse_prusti()?;
+            if let Some(_) = self.pop() {
```

Comment:
```suggestion
            if self.pop().is_some() {
```

Reply:
On second thought, given how you use this match, you might want to add something like:
```rust
fn assert_end(&mut self) -> syn::Result<()> {
    if self.pop().is_some() {
        Err(self.error_unexpected())
    } else {
        Ok(())
    }
}
```
And then call `self.assert_end()?;` instead of the conditionals.

Reply:
I think `self.pop().is_some()` is sufficiently concise that it doesn't need to be extracted into a function.

Reply:
First part addressed by c77b6d0b1638c36b81eddb1e87af37c3f1fff212